pkg/console/oidcsetup/oidcsetup.go

```diff
@@ -0,0 +1,262 @@
+	utilsub "github.com/openshift/console-operator/pkg/console/subresource/util"
+)
+
+type oidcSetupController struct {
+	operatorClient v1helpers.OperatorClient
+
+	authnLister               configv1listers.AuthenticationLister
+	crdLister                 apiexensionsv1listers.CustomResourceDefinitionLister
+	consoleOperatorLister     operatorv1listers.ConsoleLister
+	targetNSSecretsLister     corev1listers.SecretLister
+	targetNSConfigMapLister   corev1listers.ConfigMapLister
+	targetNSDeploymentsLister appsv1listers.DeploymentLister
+
+	authStatusHandler *status.AuthStatusHandler
+}
+
+func NewOIDCSetupController(
+	operatorClient v1helpers.OperatorClient,
+	authnInformer configv1informers.AuthenticationInformer,
+	authenticationClient configv1client.AuthenticationInterface,
+	consoleOperatorInformer operatorv1informers.ConsoleInformer,
+	crdInformer apiexensionsv1informers.CustomResourceDefinitionInformer,
+	targetNSsecretsInformer corev1informers.SecretInformer,
+	targetNSConfigMapInformer corev1informers.ConfigMapInformer,
+	targetNSDeploymentsInformer appsv1informers.DeploymentInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &oidcSetupController{
+		operatorClient: operatorClient,
+
+		authnLister:               authnInformer.Lister(),
+		consoleOperatorLister:     consoleOperatorInformer.Lister(),
+		crdLister:                 crdInformer.Lister(),
+		targetNSSecretsLister:     targetNSsecretsInformer.Lister(),
+		targetNSDeploymentsLister: targetNSDeploymentsInformer.Lister(),
+		targetNSConfigMapLister:   targetNSConfigMapInformer.Lister(),
+
+		authStatusHandler: status.NewAuthStatusHandler(authenticationClient, api.OpenShiftConsoleName, api.TargetNamespace, api.OpenShiftConsoleOperator),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(wait.Jitter(time.Minute, 1.0)).
+		WithFilteredEventsInformers(
+			factory.NamesFilter("authentications.config.openshift.io"),
+			crdInformer.Informer(),
+		).
+		WithInformers(
+			authnInformer.Informer(),
+			consoleOperatorInformer.Informer(),
+			targetNSsecretsInformer.Informer(),
+			targetNSDeploymentsInformer.Informer(),
+			targetNSConfigMapInformer.Informer(),
+		).
+		ToController("OIDCSetupController", recorder.WithComponentSuffix("oidc-setup-controller"))
+}
+
+func (c *oidcSetupController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	statusHandler := status.NewStatusHandler(c.operatorClient)
+
+	if shouldSync, err := c.handleManaged(ctx); err != nil {
+		return err
+	} else if !shouldSync {
+		return nil
+	}
+
+	operatorConfig, err := c.consoleOperatorLister.Get(api.ConfigResourceName)
+	if err != nil {
+		return err
+	}
+
+	authnConfig, err := c.authnLister.Get(api.ConfigResourceName)
+	if err != nil {
+		return err
+	}
+
+	if authnConfig.Spec.Type == configv1.AuthenticationTypeOIDC {
+		err = c.syncAuthTypeOIDC(ctx, syncCtx, statusHandler, operatorConfig, authnConfig)
```

Comment:
sticky.  I think I finally get this and I recommend some factorization updates.  The conditional part of this logic is "get the clientID if available in spec to be provided in status" and "if the clientID is available, determine the status of using this clientID".

I'm wondering if most of this function control loop is actually

```go
func sync(){
  return if not managed
  if !oidcClientsSchema{
    clear all conditions
    return
  }

  if .spec.type != OIDC{
    apply an empty authentication.config.openshift.io .status.oidcClients[componentNamespace=openshift-console,componentName=console]
    return
  } 

  read current authentication .spec.oidc.oidcClient[console]
  check status of rolling out the clientid 
  apply complete authentication.config.openshift.io .status.oidcClients[componentNamespace=openshift-console,componentName=console]

  which would eliminate the if-blocks-with-continue that made this difficult to read.
}
```

Reply:
I rewrote pieces of it so that this error does not make the loop finish early.